content/post/2025-02-01-lazy/index.Rmd

```diff
@@ -0,0 +1,225 @@
+bench::mark(lazy_future <- future(is.numeric(runif(n = 10000000)), lazy = TRUE))
+```
+
+If we do retrieve the value, overall the same time is spent between creating the future and our getting the value:
+
+```{r, cache=TRUE, warning=FALSE}
+bench::mark({
+  no_future <- is.numeric(runif(n = 10000000))
+  no_future
+})
+
+bench::mark({
+  eager_future <- future(is.numeric(runif(n = 10000000)))
+  value(eager_future)
+})
+
+bench::mark({
+  lazy_future <- future(is.numeric(runif(n = 10000000)), lazy = TRUE)
+  value(lazy_future)
+})
+```
+
+Therefore, the use of futures and the use of lazy evaluation are orthogonal concepts: you can use future with or without lazy evaluation.
+The future package is about *how* the value is computed (in parallel or sequentially for instance), lazy evaluation is about *when* the value is computed (right as it is defined, or only when it is needed).
+
+## Lazy as in lazy database operations
+
+In the database world, queries can be lazy: the query is like a TODO list that is only executed (computed, evaluated) when you want to access the resulting table or result.
+Making the output tangible is called *materialization*.
+
+This is vocabulary we can encounter when using:
+
+-   the [dbplyr package](https://dbplyr.tidyverse.org/), which is the dplyr back-end for databases. "All dplyr calls are evaluated lazily, generating SQL that is only sent to the database when you request the data."
+
+Slightly tweaked from the dbplyr README,
+
+```{r}
+# load packages
+library(dplyr, warn.conflicts = FALSE)
+library("dbplyr")
+
+# create the connection and refer to the table
+con <- DBI::dbConnect(RSQLite::SQLite(), ":memory:")
+copy_to(con, mtcars)
+mtcars2 <- tbl(con, "mtcars")
+
+# create the query
+summary <- mtcars2 %>% 
+  group_by(cyl) %>% 
+  summarise(mpg = mean(mpg, na.rm = TRUE)) %>% 
+  arrange(desc(mpg))
+
+# the object is lazy, the value is not computed yet
+summary
+nrow(summary)
+
+# we explictly request the data, so now it's there
+answer <- collect(summary)
+nrow(answer)
+
+```
+
+-   the [dtplyr package](https://dtplyr.tidyverse.org/index.html), which is a data.table back-end for dplyr. The ["lazy" data.table objects](https://dtplyr.tidyverse.org/reference/lazy_dt.html) "captures the intent of dplyr verbs, only actually performing computation when requested" (with `collect()` for instance). The manual also explains that this allows dtplyr to make the code more performant by simplifying the data.table calls.
+
+Slightly tweaked from dtplyr README,
+
+```{r}
+# load packages
+library(data.table)
+library(dtplyr)
+library(dplyr, warn.conflicts = FALSE)
+
+# create a “lazy” data table that tracks the operations performed on it.
+mtcars2 <- lazy_dt(mtcars)
+
+# create the query
+summary <- mtcars2 %>% 
+  filter(wt < 5) %>% 
+  mutate(l100k = 235.21 / mpg) %>% # liters / 100 km
+  group_by(cyl) %>% 
+  summarise(l100k = mean(l100k))
+
+# the object is lazy, the value is not computed yet
+summary
```

Comment:
How does it know what to put into the summary? Something clearly is already there, maybe you can add a sentence on that (possibly even just to point elsewhere for an explanation).

Reply:
I added a sentence but really I have no idea so the sentence is "here's how it looks like at this stage" :sweat_smile: